ERROR: smallint out of range occurred in ShardingSphere-Proxy PostgreSQL when inserting negative value into smallint by prepared statement

## Bug Report

### Which version of ShardingSphere did you use?

7aceb284a771d4c919920343f609615c8d2d65de

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?

ShardingSphere-Proxy PostgreSQL

### Expected behavior

Insert succeeded.

### Actual behavior

```
[ERROR] 2022-11-22 14:09:56.577 [Connection-3-ThreadExecutor] o.a.s.p.f.c.CommandExecutorTask - Exception occur: 
org.postgresql.util.PSQLException: ERROR: smallint out of range
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2676)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2366)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:356)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:490)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:408)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:181)
	at org.postgresql.jdbc.PgPreparedStatement.execute(PgPreparedStatement.java:170)
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.execute(ProxyPreparedStatement.java:44)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.execute(HikariProxyPreparedStatement.java)
	at org.apache.shardingsphere.proxy.backend.communication.jdbc.executor.callback.impl.ProxyPreparedStatementExecutorCallback.execute(ProxyPreparedStatementExecutorCallback.java:43)
	at org.apache.shardingsphere.proxy.backend.communication.jdbc.executor.callback.ProxyJDBCExecutorCallback.executeSQL(ProxyJDBCExecutorCallback.java:75)
	at org.apache.shardingsphere.proxy.backend.communication.jdbc.executor.callback.ProxyJDBCExecutorCallback.executeSQL(ProxyJDBCExecutorCallback.java:68)
	at org.apache.shardingsphere.proxy.backend.communication.jdbc.executor.callback.ProxyJDBCExecutorCallback.executeSQL(ProxyJDBCExecutorCallback.java:45)
	at org.apache.shardingsphere.infra.executor.sql.execute.engine.driver.jdbc.JDBCExecutorCallback.execute(JDBCExecutorCallback.java:90)
	at org.apache.shardingsphere.infra.executor.sql.execute.engine.driver.jdbc.JDBCExecutorCallback.execute(JDBCExecutorCallback.java:69)
	at org.apache.shardingsphere.infra.executor.kernel.ExecutorEngine.syncExecute(ExecutorEngine.java:135)
	at org.apache.shardingsphere.infra.executor.kernel.ExecutorEngine.parallelExecute(ExecutorEngine.java:131)
	at org.apache.shardingsphere.infra.executor.kernel.ExecutorEngine.execute(ExecutorEngine.java:116)
	at org.apache.shardingsphere.infra.executor.sql.execute.engine.driver.jdbc.JDBCExecutor.execute(JDBCExecutor.java:67)
	at org.apache.shardingsphere.proxy.backend.communication.jdbc.executor.ProxyJDBCExecutor.execute(ProxyJDBCExecutor.java:75)
	at org.apache.shardingsphere.proxy.backend.communication.ProxySQLExecutor.useDriverToExecute(ProxySQLExecutor.java:229)
	at org.apache.shardingsphere.proxy.backend.communication.ProxySQLExecutor.doExecute(ProxySQLExecutor.java:186)
	at org.apache.shardingsphere.proxy.backend.communication.ProxySQLExecutor.execute(ProxySQLExecutor.java:151)
	at org.apache.shardingsphere.proxy.backend.communication.jdbc.JDBCDatabaseCommunicationEngine.execute(JDBCDatabaseCommunicationEngine.java:131)
	at org.apache.shardingsphere.proxy.frontend.postgresql.command.query.extended.JDBCPortal.bind(JDBCPortal.java:101)
	at org.apache.shardingsphere.proxy.frontend.postgresql.command.query.extended.bind.PostgreSQLComBindExecutor.execute(PostgreSQLComBindExecutor.java:53)
	at org.apache.shardingsphere.proxy.frontend.postgresql.command.query.extended.PostgreSQLAggregatedCommandExecutor.execute(PostgreSQLAggregatedCommandExecutor.java:41)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.executeCommand(CommandExecutorTask.java:111)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.run(CommandExecutorTask.java:78)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

### Example codes for reproduce this issue (such as a github link).

```java
        try (Statement statement = connection.createStatement()) {
            statement.executeUpdate("drop table if exists pg_int");
            statement.executeUpdate("create table pg_int\n" +
                    "(\n" +
                    "    val_big    bigint    not null,\n" +
                    "    val_int    int       not null,\n" +
                    "    val_small  smallint  not null\n" +
                    ")");
        }
        try (PreparedStatement preparedStatement = connection.prepareStatement("insert into pg_int values (?, ?, ?)")) {
            preparedStatement.setLong(1, -1);
            preparedStatement.setInt(2, -1);
            preparedStatement.setShort(3, (short) -1);
            preparedStatement.executeUpdate();
        }
```

--- a/db-protocol/postgresql/src/main/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/bind/protocol/PostgreSQLInt2BinaryProtocolValue.java
+++ b/db-protocol/postgresql/src/main/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/bind/protocol/PostgreSQLInt2BinaryProtocolValue.java
@@ -31,7 +31,7 @@ public final class PostgreSQLInt2BinaryProtocolValue implements PostgreSQLBinary
     
     @Override
     public Object read(final PostgreSQLPacketPayload payload, final int parameterValueLength) {
-        return payload.readInt2();
+        return payload.getByteBuf().readShort();
     }
     
     @Override

--- a/db-protocol/postgresql/src/test/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/bind/protocol/PostgreSQLInt2BinaryProtocolValueTest.java
+++ b/db-protocol/postgresql/src/test/java/org/apache/shardingsphere/db/protocol/postgresql/packet/command/query/extended/bind/protocol/PostgreSQLInt2BinaryProtocolValueTest.java
@@ -17,30 +17,40 @@
 
 package org.apache.shardingsphere.db.protocol.postgresql.packet.command.query.extended.bind.protocol;
 
+import io.netty.buffer.Unpooled;
 import org.apache.shardingsphere.db.protocol.postgresql.payload.PostgreSQLPacketPayload;
 import org.junit.Test;
-import org.junit.runner.RunWith;
-import org.mockito.Mock;
-import org.mockito.junit.MockitoJUnitRunner;
+
+import java.nio.charset.StandardCharsets;
 
 import static org.hamcrest.CoreMatchers.is;
 import static org.hamcrest.MatcherAssert.assertThat;
-import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.when;
 
-@RunWith(MockitoJUnitRunner.class)
 public final class PostgreSQLInt2BinaryProtocolValueTest {
     
-    @Mock
-    private PostgreSQLPacketPayload payload;
+    @Test
+    public void assertGetColumnLength() {
+        assertThat(new PostgreSQLInt2BinaryProtocolValue().getColumnLength(null), is(2));
+    }
     
     @Test
-    public void assertNewInstance() {
+    public void assertRead() {
+        byte[] data = {(byte) 0x80, (byte) 0x00, (byte) 0xFF, (byte) 0xFF, (byte) 0x7F, (byte) 0xFF};
         PostgreSQLInt2BinaryProtocolValue actual = new PostgreSQLInt2BinaryProtocolValue();
-        assertThat(actual.getColumnLength(null), is(2));
-        when(payload.readInt2()).thenReturn(1);
-        assertThat(actual.read(payload, 2), is(1));
-        actual.write(payload, 1);
-        verify(payload).writeInt2(1);
+        PostgreSQLPacketPayload payload = new PostgreSQLPacketPayload(Unpooled.wrappedBuffer(data), StandardCharsets.UTF_8);
+        assertThat(actual.read(payload, 2), is(Short.MIN_VALUE));
+        assertThat(actual.read(payload, 2), is((short) -1));
+        assertThat(actual.read(payload, 2), is(Short.MAX_VALUE));
+    }
+    
+    @Test
+    public void assertWrite() {
+        byte[] actualData = new byte[6];
+        PostgreSQLPacketPayload payload = new PostgreSQLPacketPayload(Unpooled.wrappedBuffer(actualData).writerIndex(0), StandardCharsets.UTF_8);
+        PostgreSQLInt2BinaryProtocolValue actual = new PostgreSQLInt2BinaryProtocolValue();
+        actual.write(payload, (int) Short.MIN_VALUE);
+        actual.write(payload, -1);
+        actual.write(payload, (int) Short.MAX_VALUE);
+        assertThat(actualData, is(new byte[]{(byte) 0x80, (byte) 0x00, (byte) 0xFF, (byte) 0xFF, (byte) 0x7F, (byte) 0xFF}));
     }
 }

--- a/test/integration-test/test-suite/src/test/resources/cases/dml/dataset/tbl/insert_min_values_into_single_table_integer.xml
+++ b/test/integration-test/test-suite/src/test/resources/cases/dml/dataset/tbl/insert_min_values_into_single_table_integer.xml
@@ -24,5 +24,5 @@
         <column name="col_smallint" type="numeric" />
         <column name="col_tinyint" type="numeric" />
     </metadata>
-    <row data-node="tbl.t_single_table_integer" values="2, -9223372036854775808, -2147483648, -8388608, 0, 0" />
+    <row data-node="tbl.t_single_table_integer" values="2, -9223372036854775808, -2147483648, -8388608, -32768, -128" />
 </dataset>

--- a/test/integration-test/test-suite/src/test/resources/cases/dml/dataset/tbl/insert_negative_values_into_single_table_integer.xml
+++ b/test/integration-test/test-suite/src/test/resources/cases/dml/dataset/tbl/insert_negative_values_into_single_table_integer.xml
@@ -24,5 +24,5 @@
         <column name="col_smallint" type="numeric" />
         <column name="col_tinyint" type="numeric" />
     </metadata>
-    <row data-node="tbl.t_single_table_integer" values="3, -1, -1, -1, 0, 0" />
+    <row data-node="tbl.t_single_table_integer" values="3, -1, -1, -1, -1, -1" />
 </dataset>

--- a/test/integration-test/test-suite/src/test/resources/cases/dml/dml-integration-test-cases.xml
+++ b/test/integration-test/test-suite/src/test/resources/cases/dml/dml-integration-test-cases.xml
@@ -299,8 +299,7 @@
     </test-case>
     <test-case sql="INSERT INTO t_single_table_integer (id, col_bigint, col_int, col_mediumint, col_smallint, col_tinyint) values (?, ?, ?, ?, ?, ?)" db-types="PostgreSQL,openGauss" scenario-types="tbl">
         <assertion parameters="1:int, 9223372036854775807:long, 2147483647:int, 8388607:int, 32767:smallint, 127:tinyint" expected-data-file="insert_max_values_into_single_table_integer.xml" />
-        <!-- TODO Support negative value of smallint(int2). https://github.com/apache/shardingsphere/issues/22334 -->
-        <assertion parameters="2:int, -9223372036854775808:long, -2147483648:int, -8388608:int, 0:smallint, 0:tinyint" expected-data-file="insert_min_values_into_single_table_integer.xml" />
-        <assertion parameters="3:int, -1:long, -1:int, -1:int, 0:smallint, 0:tinyint" expected-data-file="insert_negative_values_into_single_table_integer.xml" />
+        <assertion parameters="2:int, -9223372036854775808:long, -2147483648:int, -8388608:int, -32768:smallint, -128:tinyint" expected-data-file="insert_min_values_into_single_table_integer.xml" />
+        <assertion parameters="3:int, -1:long, -1:int, -1:int, -1:smallint, -1:tinyint" expected-data-file="insert_negative_values_into_single_table_integer.xml" />
     </test-case>
 </integration-test-cases>

